gateway-sp-comms/src/management_switch/location_map.rs

```diff
@@ -0,0 +1,461 @@
+    pub(super) fn new_raw(
+        location_name: String,
+        port_to_id: HashMap<SwitchPort, SpIdentifier>,
+    ) -> Self {
+        let mut id_to_port = HashMap::with_capacity(port_to_id.len());
+        for (&port, &id) in port_to_id.iter() {
+            id_to_port.insert(id, port);
+        }
+        Self { location_name, port_to_id, id_to_port }
+    }
+
+    pub(super) async fn run_discovery(
+        communicator: Arc<crate::Communicator>,
+        config: LocationConfig,
+        ports: HashMap<SwitchPort, SwitchPortConfig>,
+        sockets: Arc<HashMap<SwitchPort, UdpSocket>>,
+        recv_handler: Arc<RecvHandler>,
+        deadline: Instant,
+        log: &Logger,
+    ) -> Result<Self, StartupError> {
+        // Spawn a task that will send discovery packets on every switch port
+        // until it hears back from all SPs with exponential backoff to avoid
+        // slamming the network; we expect some ports to never resolve (e.g., if
+        // the cubby at the other end is not populated and therefore there is no
+        // SP to respond).
+        let location_determination = config.determination;
+        let (refined_locations_tx, refined_locations) = mpsc::channel(8);
+        {
+            let log = log.clone();
+            let ports = ports.clone();
+            tokio::spawn(async move {
+                discover_sps(
+                    &communicator,
+                    &sockets,
+                    ports,
+                    &recv_handler,
+                    location_determination,
+                    refined_locations_tx,
+                    &log,
+                )
+                .await;
+            });
+        }
+
+        // Collect responses and solve for a single location
+        let location = match tokio::time::timeout_at(
+            deadline,
+            resolve_location(
+                config.names,
+                ReceiverStream::new(refined_locations),
+                log,
+            ),
+        )
+        .await
+        {
+            Ok(Ok(location)) => location,
+            Ok(Err(err)) => return Err(err),
+            Err(_) => {
+                return Err(StartupError::DiscoveryFailed {
+                    reason: String::from("timeout"),
+                })
+            }
+        };
+
+        // based on the resolved location and the input configuration, build the
+        // map of port <-> logical ID
+        let mut port_to_id = HashMap::with_capacity(ports.len());
+        let mut id_to_port = HashMap::with_capacity(ports.len());
+        for (port, mut port_config) in ports {
+            let id =
+                port_config.location.remove(&location).ok_or_else(|| {
+                    StartupError::InvalidConfig {
+                        reason: format!(
+                            "port {} has no entry for determined location {:?}",
+                            port.0, location,
+                        ),
+                    }
+                })?;
```

Comment:
might we want to see all such errors rather that just the first? Or perhaps log out all mismatches and return an error saying that the config was bad?

Reply:
I considered doing some kind of full-config-file-validation up front; that seems better than failing here _after_ we've already talked to some SPs. I'll do that and log all errors.

Reply:
Added config validation in 674ffc27; it's a little long/boilerplatey but I think that's probably fine in exchange for getting a complete list of errors.